### DISCUSSION TITLE
How to set routes that only they can access Context variables

### DISCUSSION THREAD

**Discussion by PatricioPoncini** [honojs/hono#1987] [General]
Hi, I'm wondering if there's a way to configure routes in a manner that restricts access to certain Context variables. For instance:

```javascript
interface Env {
  Variables: { foo: 'bar' };
}

const route = new Hono<Env>().basePath('/');
```

In this example, using c.get(foo) within this route retrieves the value of the 'foo' key. However, this implies that the 'foo' key is available across all routes that are typed with the Env interface. How can I define a route to exclusively have the Env interface and another variable that can only be accessed from that specific route?

Because if I want to extend, for example here, it will give an error in the other routes since they are typed with Env, and this one is also, but it has the addition that I'm adding something exclusive to it.
```javascript
const route = new Hono<Env & { Variables: { key: 'value' } }>().basePath('/');
```

**Comment by yusukebe** [comment]
Hi @PatricioPoncini 

This is a hot topic for me too. We are currently considering the feature to add another Env type to only a specific route. PR: #1980

```ts
type Env = {
  Variables: {
    foo: string
  }
}

const app = new Hono<Env>()

app.get<{
  Variables: {
    bar: string
  }
}>('/', (c) => {
  const foo = c.var.foo // foo is string
  const bar = c.var.bar // bar is string
  return c.json(0)
})
```

Does this solve your problem?

**Reply by beowulf11** [reply]
Are there any plans or ways to achieve this functionality?

**Reply by PatricioPoncini** [reply]
@beowulf11 Yes, this solution is effective. Try creating a new Hono project and applying this:
```ts
import { Hono } from "hono";

interface FooInterface {
  Variables: { foo: string };
}

interface BarInterface {
  Variables: { bar: string };
}

// If you only want one of them, you can use just one interface
const app = new Hono<FooInterface & BarInterface>();

app.get("/", (c) => {
  c.set("foo", "bar");
  c.set("bar", "foo");
  const fooValue = c.get("foo");
  const barValue = c.get("bar");
  return c.json({ fooValue, barValue });
});

export default app;
```

If you want to make a curl request to that, you will get these return values:
```bash
> curl http://localhost:3000
{"barValue":"foo","fooValue":"bar"}
```
And this is how you can set it up with 2 different interfaces. I hope it helps! :smile:

**Reply by beowulf11** [reply]
Thanks for the reply! Just so I understand the idea is to create a new instance of `Hono` when there is a need to add Variables only to a specific route.

I guess that will work, I am not sure if this won't break or result in some issues when I have multiple apps for a specific sub-router. But I guess if there isn't name collision it should work fine.

**Reply by PatricioPoncini** [reply]
Unless there is a collision between the types of Hono instances, there shouldn't be any problem in my experience.
